src/celeritas/decay/interactor/MuDecayInteractor.hh

```diff
@@ -0,0 +1,299 @@
+
+    // Define decay channel
+    auto const& ids = shared_.ids;
+    bool muon = (particle.particle_id() == ids.mu_minus) ? true : false;
+    secondary_id_[0] = muon ? ids.electron : ids.positron;
+    secondary_id_[1] = muon ? ids.anti_electron_neutrino
+                            : ids.electron_neutrino;
+    secondary_id_[2] = muon ? ids.muon_neutrino : ids.anti_muon_neutrino;
+
+    // Set up muon four vector to boost decay to the lab-frame
+    Real3 inc_momentum = inc_direction_ * inc_energy_.value();
+    inc_fourvec_ = {
+        inc_momentum,
+        std::sqrt(ipow<2>(norm(inc_momentum)) + ipow<2>(shared_.muon_mass))};
+
+    // Sampling constants
+    sample_max_ = real_type{1}
+                  + ipow<2>(shared_.electron_mass) / ipow<2>(shared_.muon_mass);
+
+    // Geant4 physics manual defines E_{max} = m_\mu / 2, while the source code
+    // (since v10.2.0 at least) defines E_{max} = m_\mu / 2 - m_e . The source
+    // code implementation leads to a total CM energy of ~104.6 MeV instead of
+    // the expected 105.7 MeV (muon mass), which is achieved by using the
+    // physics manual definition
+    max_energy_ = 0.5 * shared_.muon_mass;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Sample a muon decay via the
+ * \f[
+ * \mu^- \longrightarrow e^- \overline{\nu}_e \nu_\mu
+ * \f]
+ * or
+ * \f[
+ * \mu^+ \longrightarrow e^+ \nu_e \overline{\nu}_\mu
+ * \f]
+ * channel, with a branching ratio of 100%.
+ *
+ * The decay is sampled at the muon's reference frame (based on
+ * \c G4MuonDecayChannel::DecayIt ) and the resulting four vector is boosted to
+ * the lab frame using the muon's incident energy and direction.
+ *
+ * \note Neutrinos are assumed to have zero mass.
+ */
+template<class Engine>
+CELER_FUNCTION Interaction MuDecayInteractor::operator()(Engine& rng)
+{
+    // Allocate secondaries
+    Secondary* secondaries = allocate_(3);
+    if (secondaries == nullptr)
+    {
+        // Failed to allocate secondaries
+        return Interaction::from_failure();
+    }
+
+    real_type electron_energy_frac;
+    real_type electron_nu_energy_frac;
+    size_type const max_iter{1000};  // \todo Why?
+
+    for ([[maybe_unused]] auto i : range(max_iter))
+    {
+        for ([[maybe_unused]] auto j : range(max_iter))
+        {
+            electron_nu_energy_frac = sample_max_ * generate_canonical(rng);
+            if (generate_canonical(rng)
+                <= electron_nu_energy_frac
+                       * (real_type{1} - electron_nu_energy_frac))
+            {
+                break;
+            }
+            electron_nu_energy_frac = sample_max_;
+        }
+
+        electron_energy_frac = generate_canonical(rng);
+        if (electron_nu_energy_frac >= real_type{1} - electron_energy_frac)
+        {
+            break;
+        }
+        electron_nu_energy_frac = real_type{1} - electron_energy_frac;
+    }
```

Comment:
Yep. If you see the top plot on my comment just after the pull request, it shows that the number of iterations on these two `for` loops are way below that `max_iter` defined in G4. I am happy to convert this to `do while` if everyone is fine with it. My worry was that G4 used the for loops because of some very rare cases that the sampling got stuck for too long, so they just sample the max electron energy on those cases.